Sources/NIOSSL/SSLContext.swift

```diff
@@ -112,6 +112,106 @@ private func alpnCallback(ssl: OpaquePointer?,
+    
+    guard let serverCallback = parentSwiftContext.pskServerConfigurationCallback,
+          let unwrappedIdentity = identity, // Incoming identity
+          let strIdentity = String(validatingUTF8: unwrappedIdentity),
+          let configurationHint = parentSwiftContext.configuration.pskHint,
+          let outputPSK = psk // Output PSK key.
+          else {
+        return 0
+    }
+    
+    // Take the hint and the possible identity and pass them down to the callback to get associated PSK from callback
+    guard let contextPSKIdentityCallback = try? serverCallback(configurationHint, strIdentity) else {
+        return 0
+    }
+    let serverPSK = contextPSKIdentityCallback.key // From the callback
+    
+    // Make sure the key is returned by the callback and it is of proper length, otherwise, fail.
+    if serverPSK.isEmpty || serverPSK.count > max_psk_len {
+        return 0
+    }
+    let _ = serverPSK.withUnsafeBytes { (body: UnsafeRawBufferPointer) -> Void in
+        memcpy(outputPSK, body.baseAddress!, body.count)
+    }
+    return UInt32(serverPSK.count)
+}
+
+/// PSK Callback for the client side context.
+private func clientPSKCallback(ssl: OpaquePointer?,
+                                 hint: UnsafePointer<CChar>?,
+                                 identity: UnsafeMutablePointer<CChar>?,
+                                 max_identity_len: UInt32,
+                                 psk: UnsafeMutablePointer<UInt8>?,
+                                 max_psk_len: UInt32) -> UInt32 {
+    
+    guard let ssl = ssl else { return 0 }
+    // Initial implementation only supports TLS 1.2 due API support exposed from BoringSSL.
+    // TODO (meaton) add TLS 1.3 support when available.
+    
+    // We want to take the SSL pointer and extract the parent Swift object.
+    let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+    let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+    let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+    
+    guard let clientCallback = parentSwiftContext.pskClientConfigurationCallback,
+          let unwrappedIdentity = identity, // Output identity that will be later be mapped from client callback
+          let outputPSK = psk // Output PSK key that will later be mapped from client callback
+          else {
+        return 0
+    }
+    
+    // If set, build out a hint to pass into the client callback.
+    guard let clientHint = hint,
+          let derivedHint = String(validatingUTF8: clientHint) else {
+        return 0
+    }
+    // Take the hint and pass it down to the callback to get associated PSK from callback
+    guard let pskIdentityCallback = try? clientCallback(derivedHint) else {
+        return 0
+    }
+    let clientPSK = pskIdentityCallback.key // Key from the callback
+    let clientIdentity = pskIdentityCallback.identity
+    
+    if clientIdentity.isEmpty {
```

Comment:
This check is redundant with the one below it.

Reply:
Addressed in [f961f4a](https://github.com/apple/swift-nio-ssl/pull/369/commits/f961f4a6d34a6f5f40bc79b33e38e90ae76a12e3)